cmd/insights/commands/upload.go

```diff
@@ -40,23 +42,45 @@ func installUploadCmd(app *App) {
 	uploadCmd.Flags().UintVar(&app.config.Upload.MinAge, "min-age", defaultMinAge, "the minimum age (in seconds) of a report before the uploader will attempt to upload it")
 	uploadCmd.Flags().BoolVarP(&app.config.Upload.Force, "force", "f", false, "force an upload, ignoring min age and clashes between the collected file and a file in the uploaded folder, replacing the clashing uploaded report if it exists")
 	uploadCmd.Flags().BoolVarP(&app.config.Upload.DryRun, "dry-run", "d", false, "go through the motions of doing an upload, but do not communicate with the server or send the payload")
+	uploadCmd.Flags().BoolVarP(&app.config.Upload.BackoffRetry, "backoff-retry", "b", false, "enable exponential backoff retry for failed uploads")
 
 	app.cmd.AddCommand(uploadCmd)
 }
 
 func (a App) uploadRun() error {
 	cm := consent.New(a.config.consentDir)
 
+	var wg sync.WaitGroup
+	uploaders := make(map[string]uploader.Uploader)
 	for _, source := range a.config.Upload.Sources {
-		u, err := a.newUploader(cm, a.config.insightsDir, source, a.config.Upload.MinAge, a.config.Upload.DryRun)
+		u, err := a.newUploader(cm, a.config.insightsDir, source, a.config.Upload.MinAge, a.config.Upload.DryRun, a.config.Upload.BackoffRetry)
 		if err != nil {
 			return fmt.Errorf("failed to create uploader for source %s: %v", source, err)
 		}
-
-		if err := u.Upload(a.config.Upload.Force); err != nil {
-			return fmt.Errorf("failed to upload reports for source %s: %v", source, err)
-		}
+		uploaders[source] = u
 	}
 
-	return nil
+	var uploadError error
+	mu := &sync.Mutex{}
+	for s, u := range uploaders {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			var err error
+			if a.config.Upload.BackoffRetry {
+				err = u.BackoffUpload(a.config.Upload.Force)
+			} else {
+				err = u.Upload(a.config.Upload.Force)
+			}
+
+			if err != nil {
+				errMsg := fmt.Errorf("failed to upload reports for source %s: %v", s, err)
+				mu.Lock()
+				uploadError = errors.Join(uploadError, errMsg)
+				mu.Unlock()
```

Comment:
```suggestion
				mu.Lock()
				defer mu.Unlock()
				uploadError = errors.Join(uploadError, errMsg)
```

Not that errors.Join will panic, but if it did, you'd never unlock. Everytime I acquire a resource I must release, I place the release call in a defer statement right below the acquisition, this way I don't need to think whether the code below would panic or not.